core/sizes.cpp

```diff
@@ -2,81 +2,175 @@
 
-scipp::index &Sizes::operator[](const Dim dim) { return at(dim); }
+template <class Key, class Value, int16_t Capacity>
+bool small_stable_map<Key, Value, Capacity>::operator==(
+    const small_stable_map &other) const noexcept {
+  if (size() != other.size())
+    return false;
+  for (const auto &key : *this)
+    if (!other.contains(key) || at(key) != other.at(key))
+      return false;
+  return true;
+}
+
+template <class Key, class Value, int16_t Capacity>
+bool small_stable_map<Key, Value, Capacity>::operator!=(
+    const small_stable_map &other) const noexcept {
+  return !operator==(other);
+}
+
+template <class Key, class Value, int16_t Capacity>
+typename std::array<Key, Capacity>::const_iterator
+small_stable_map<Key, Value, Capacity>::find(const Key &key) const {
+  return std::find(begin(), end(), key);
+}
 
-scipp::index Sizes::at(const Dim dim) const {
-  scipp::expect::contains(*this, dim);
-  return m_sizes.at(dim);
+template <class Key, class Value, int16_t Capacity>
+bool small_stable_map<Key, Value, Capacity>::contains(const Key &key) const {
+  return find(key) != end();
 }
 
-scipp::index &Sizes::at(const Dim dim) {
-  scipp::expect::contains(*this, dim);
-  return m_sizes.at(dim);
+template <class Key, class Value, int16_t Capacity>
+scipp::index
+small_stable_map<Key, Value, Capacity>::index(const Key &key) const {
+  const auto it = find(key);
+  if (it == end())
+    throw_dimension_not_found_error(*this, key);
+  return std::distance(begin(), it);
 }
 
+template <class Key, class Value, int16_t Capacity>
+const Value &
+small_stable_map<Key, Value, Capacity>::operator[](const Key &key) const {
+  return at(key);
+}
+
+template <class Key, class Value, int16_t Capacity>
+const Value &small_stable_map<Key, Value, Capacity>::at(const Key &key) const {
+  return m_values[index(key)];
+}
+
+template <class Key, class Value, int16_t Capacity>
+void small_stable_map<Key, Value, Capacity>::assign(const Key &key,
+                                                    const Value &value) {
+  m_values[index(key)] = value;
+}
+
+template <class Key, class Value, int16_t Capacity>
+void small_stable_map<Key, Value, Capacity>::insert_left(const Key &key,
+                                                         const Value &value) {
+  expectUnique(*this, key);
+  expectExtendable(*this);
+  for (scipp::index i = size(); i > 0; --i) {
+    m_keys[i] = m_keys[i - 1];
+    m_values[i] = m_values[i - 1];
+  }
+  m_keys[0] = key;
+  m_values[0] = value;
+  m_size++;
+}
+
+template <class Key, class Value, int16_t Capacity>
+void small_stable_map<Key, Value, Capacity>::insert_right(const Key &key,
+                                                          const Value &value) {
+  expectUnique(*this, key);
+  expectExtendable(*this);
+  m_keys[m_size] = key;
+  m_values[m_size] = value;
+  m_size++;
+}
+
+template <class Key, class Value, int16_t Capacity>
+void small_stable_map<Key, Value, Capacity>::erase(const Key &key) {
```

Comment:
Shouldn't this be noexcept? Ah maybe not since `index` can throw. I was going to suggest that it be safer to implement the erase as a copy/swap type operation, however I think anything that can throw will do before the modifications start.

Should we return for operations such as this?

Reply:
If key does not exist it throws.